common/aggregate/moving_window_average.go

```diff
@@ -0,0 +1,115 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package aggregate
+
+import (
+	"container/ring"
+	"sync"
+	"time"
+)
+
+type (
+	MovingWindowAverage interface {
+		Record(val int64)
+		Average() float64
+	}
+
+	timestampedData struct {
+		value     int64
+		timestamp time.Time
+	}
+
+	MovingWindowAvgImpl struct {
+		sync.RWMutex
+		windowSize    time.Duration
+		maxBufferSize int
+		head          *ring.Ring
+		tail          *ring.Ring
+		sum           int64
+		count         int64
+	}
+)
+
+func NewMovingWindowAvgImpl(
+	windowSize time.Duration,
+	maxBufferSize int,
+) *MovingWindowAvgImpl {
+	buffer := ring.New(maxBufferSize)
+	return &MovingWindowAvgImpl{
+		windowSize:    windowSize,
+		maxBufferSize: maxBufferSize,
+		head:          buffer,
+		tail:          buffer,
+	}
+}
+
+func (a *MovingWindowAvgImpl) Record(val int64) {
+	a.Lock()
+	defer a.Unlock()
+
+	if a.count == int64(a.maxBufferSize) {
+		a.expireOneLocked()
+	}
+
+	a.tail.Value = timestampedData{value: val, timestamp: time.Now()}
+	a.tail = a.tail.Next()
+
+	a.sum += val
+	a.count++
+}
```

Comment:
if user call `Record` in a loop, then once the ring is full, the oldest value will be thrown away, even if timestamp has not expired?

Reply:
Yes, the intent is not to guarantee that we capture every observation in the time window, just enough to make sure we get an accurate picture of persistence health. Basically, we are taking all the observations within the time window, or the most recent maxBufferSize observations, whichever comes first.

I didn't setup the git branches most effectively, so most of the discussion around the metric collection piece is in this PR: https://github.com/temporalio/temporal/pull/4365

Reply:
i would say it will be harder for dev to estimate / evaluate the behavior if suddenly the # of invocations increased.

is n number of time buckets (e.g. circular time buckets) a better idea or not?

Reply:
Why do you think it will be harder to estimate? If the number of invocations suddenly increases and persistence does not handle the load well, we would expect to see an increase in latency and/or error rates.

If we run into a situation where the buffer is constantly overflowing and the oldest observations are getting dropped, I'm not sure that is a significant problem. I would expect that both healthy and unhealthy observations would get dropped in the same ratio, so getting the most recent average should still give an accurate indication of persistence health. Of course we would still want to increase the buffer size if it is constantly overflowing; I just don't thinking losing a handful of observations will affect the average much.